File: 백준/Gold/10800. 컬러볼/컬러볼.py
import sys
input = sys.stdin.readline
N = int(input())
balls = []

for i in range(N):
    color, size = map(int, input().split())
    balls.append((size, color, i))

balls.sort(key=lambda x : (x[0], x[1], x[2]))
color_count = [0] * 200001
total = 0
j = 0
ans = [0] * (N)

for i in range(N):
    while balls[j][0] < balls[i][0]: # 내 공 크기보다 사이즈가 작아야 잡을 수 있음.
        color_count[balls[j][1]] += balls[j][0] # 각 색깔에 해당하는 총 공의 크기
        total += balls[j][0]
        j += 1
    # 누적합 계산
    # 나보다 작은 공들의 크기 총합에서 나랑 색 같은 것들은 빼주기
    ans[balls[i][2]] = total - color_count[balls[i][1]]

for i in range(N):
    print(ans[i])

File: 백준/Gold/10800. 컬러볼/컬러볼.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;
import java.util.Arrays;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());
        Ball[] balls = new Ball[N];
        for (int i = 0; i < N; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int color = Integer.parseInt(st.nextToken());
            int size = Integer.parseInt(st.nextToken());
            balls[i] = new Ball(size, color, i);
        }
        Arrays.sort(balls);

        int[] answer = new int[N];
        int[] colorCnt = new int[N+1];
        int j = 0;
        int total = 0;
        for (int i = 0; i < N; i++) {
            Ball cur = balls[i];
            while (balls[j].size < cur.size) {
                colorCnt[balls[j].color] += balls[j].size;
                total += balls[j].size;
                j++;
            }
            answer[cur.idx] = total - colorCnt[cur.color];
        }
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < N; i++) {
            sb.append(answer[i]).append("\n");
        }
        System.out.println(sb);

    }
    static class Ball implements Comparable<Ball> {
        int size, color, idx;

        public Ball(int size, int color, int idx) {
            this.size = size;
            this.color = color;
            this.idx = idx;
        }

        @Override
        public int compareTo(Ball b) {
            if (this.size == b.size) {
                return this.color - b.color;
            }
            return this.size - b.size;
        }
    }
}
